Core/ExpectedException/ExpectedExceptionModel.cs

```diff
@@ -0,0 +1,179 @@
+    {
+      var attributeSyntax = attribute.ApplicationSyntaxReference.GetSyntax() as AttributeSyntax
+                            ?? throw new ArgumentException ($"{nameof(attribute)} does not support a syntax tree.");
+
+      var attributeArguments = attributeSyntax.ArgumentList?.Arguments;
+
+      return attributeArguments == null
+          ? CreateNoArgExpectedExceptionModel (attribute)
+          : CreateWithArgsExpectedExceptionModel (attribute, attributeArguments.Value);
+    }
+
+    private static ExpectedExceptionModel CreateWithArgsExpectedExceptionModel (
+        AttributeData attribute,
+        SeparatedSyntaxList<AttributeArgumentSyntax> attributeArguments)
+    {
+      var builder = new ExpectedExceptionModelBuilder();
+      foreach (var attributeArgument in attributeArguments)
+      {
+        var value = attributeArgument.Expression;
+
+        var namedArgumentName = attributeArgument.NameColon?.Name
+                                ?? attributeArgument.NameEquals?.Name;
+
+        builder = namedArgumentName?.ToString() switch
+        {
+            "UserMessage" => builder.WithUserMessage (value),
+            "ExpectedException" => builder.WithExceptionType (value),
+            "ExpectedMessage" => builder.WithExpectedMessage (value),
+            "MatchType" => builder.WithMatchType (value),
+            "ExpectedExceptionName" => builder.WithExceptionName (value),
+            null => builder.WithExceptionTypeOrName (value),
+            _ => builder
+        };
+      }
+
+      return builder.Build (attribute);
+    }
+
+    private static ExpectedExceptionModel CreateNoArgExpectedExceptionModel (AttributeData attribute)
+    {
+      var builder = new ExpectedExceptionModelBuilder();
+      return builder.WithExceptionType (TypeOfExpression (IdentifierName (nameof(Exception))))
+          .Build (attribute);
+    }
+
+    private ExpressionSyntax CreateWithMessage (ExpressionSyntax expression, string baseIndent)
+    {
+      if (ExpectedMessage == null) return expression;
+
+      var indent = new string (Formatting.IndentationCharacter, Formatting.IndentationSize);
+      var withMessage = MemberAccess (
+          expression.WithTrailingTrivia (Whitespace (Formatting.NewLine + baseIndent + indent + indent)),
+          "With",
+          "Message");
+
+      var matchTypeFunctionName = MatchType?.ToFullString() switch
+      {
+          "MessageMatch.Contains" => "Contains",
+          "MessageMatch.StartsWith" => "StartsWith",
+          "MessageMatch.Regex" => "Matches",
+          "MessageMatch.Exact" => "EqualTo",
+          null => "EqualTo",
+          _ => "UNKNOWN_MATCH_TYPE"
+      };
+
+      var withMessageMatchType = MemberAccess (withMessage, matchTypeFunctionName);
+
+      if (EndsOnSameLine (ExpectedMessage)
+          && !IsMessageTooLong (ExpectedMessage))
+        return CreateSingleLineMessage (withMessageMatchType, ExpectedMessage, baseIndent + indent + indent + indent + indent);

```

Comment:
parens if the condition or the block is multiline. if one block has parens, all blocks get parens.